amethyst_renderer/src/lib.rs

```diff
@@ -0,0 +1,321 @@
+extern crate gfx_device_dx11;
+#[cfg(all(feature = "d3d11", target_os = "windows"))]
+extern crate gfx_window_dxgi;
+
+#[cfg(all(feature = "metal", target_os = "macos"))]
+extern crate gfx_device_metal;
+#[cfg(all(feature = "metal", target_os = "macos"))]
+extern crate gfx_window_metal;
+
+#[cfg(feature = "opengl")]
+extern crate gfx_device_gl;
+#[cfg(feature = "opengl")]
+extern crate gfx_window_glutin;
+#[cfg(feature = "opengl")]
+extern crate glutin;
+
+#[cfg(feature = "vulkan")]
+extern crate gfx_device_vulkan;
+#[cfg(feature = "vulkan")]
+extern crate gfx_window_vulkan;
+
+pub use cam::{Camera, Projection};
+pub use color::Rgba;
+pub use error::{Error, Result};
+pub use light::Light;
+pub use mesh::{Mesh, MeshBuilder};
+pub use mtl::{Material, MaterialBuilder};
+pub use pipe::{Pipeline, PipelineBuilder, Stage, Target};
+pub use scene::{Model, Scene};
+pub use tex::{Texture, TextureBuilder};
+pub use vertex::VertexFormat;
+
+use pipe::{ColorBuffer, DepthBuffer};
+use std::sync::Arc;
+use std::time::Duration;
+use types::{ColorFormat, DepthFormat, Encoder, Factory, Window};
+use winit::WindowBuilder;
+
+#[cfg(feature = "opengl")]
+use glutin::EventsLoop;
+#[cfg(not(feature = "opengl"))]
+use winit::EventsLoop;
+
+pub mod light;
+pub mod pass;
+pub mod prelude;
+pub mod pipe;
+pub mod vertex;
+
+mod cam;
+mod color;
+mod error;
+mod mesh;
+mod mtl;
+mod scene;
+mod tex;
+mod types;
+
+/// Generic renderer.
+pub struct Renderer {
+    device: types::Device,
+    encoders: Vec<Encoder>,
+    factory: Factory,
+    main_target: Arc<Target>,
+    pool: Arc<rayon::ThreadPool>,
+    window: Window,
+}
+
+impl Renderer {
+    /// Creates a new renderer with default window settings.
+    pub fn new(el: &EventsLoop) -> Result<Renderer> {
+        Renderer::build(el).finish()
+    }
+
+    #[allow(missing_docs)]
+    pub fn build(el: &EventsLoop) -> RendererBuilder {
+        RendererBuilder::new(el)
+    }
+
+    /// Builds a new mesh from the given vertices.
+    pub fn create_mesh(&mut self, mb: MeshBuilder) -> Result<Mesh> {
+        mb.finish(&mut self.factory)
+    }
+
+    /// Builds a new renderer pipeline.
+    pub fn create_pipe(&mut self, pb: PipelineBuilder) -> Result<Pipeline> {
+        pb.finish(&mut self.factory, &self.main_target)
+    }
+
+    /// Builds a new texture resource.
+    pub fn create_texture(&mut self, tb: TextureBuilder) -> Result<Texture> {
+        tb.finish(&mut self.factory)
+    }
+
+    /// Draws a scene with the given pipeline.
+    pub fn draw(&mut self, scene: &Scene, pipe: &Pipeline, _delta: Duration) {
+        use gfx::Device;
+        use rayon::prelude::*;
+
+        for stage in pipe.stages() {
+            assert_eq!(self.encoders.len(), num_cpus::get());
+            let encoders = self.encoders.as_mut_slice();
+            if stage.is_enabled() {
+                self.pool.install(|| {
+                    scene.par_iter_models()
+                        .zip(encoders)
```

Comment:
This code draws only `num_cpus::get()` models.
This is what (I think) it should look like.
```rust
for stage in pipe.stages() {
    if stage.is_enabled() {
        let encoders = self.encoders.as_mut_slice();
        let encoders_count = encoders.len();
        self.pool.install(|| {
            let mut enc_par_iter = encoders.into_par_iter();
            let mut mod_par_iter = scene.par_chunks_models(encoders_count);
            assert_eq!(enc_par_iter.len(), mod_par_iter.len());                        
            enc_par_iter.zip(mod_par_iter)
                .for_each(|(enc, models)| for model in models { stage.apply(enc, model, scene) });
        });
        for enc in self.encoders.iter_mut() {
            enc.flush(&mut self.device);
        }
    }
}

// in impl Scene
pub fn par_chunks_models(&self, count: usize) -> rayon::slice::Chunks<Model> {
    use rayon::prelude::*;
    let size = self.models.len();
    self.models.par_chunks(((size - 1) / count) + 1)
}
```